### ISSUE TITLE
[9.0] hr_recruitment migration raises IntegrityError

### ISSUE THREAD

**Issue by metaminux** [OCA/OpenUpgrade#1752]
Impacted versions: Migrate from 8.0 to 9.0

Steps to reproduce:
Migrate a database in v8 with modules 'hr_recruitment' installed and hr_recruitment_source records exist with id > 5 ("utm" addon is defining 5 sources)

Current behavior:
```
2019-02-15 20:19:09,985 628 INFO database_8_migrated openerp.sql_db: bad query: ALTER TABLE "hr_applicant" ADD FOREIGN KEY ("source_id") REFERENCES "utm_source" ON DELETE set null
2019-02-15 20:19:09,986 628 ERROR database_8_migrated openerp.modules.registry: Failed to load registry
Traceback (most recent call last):
  File "/var/tmp/openupgrade/9.0/server/openerp/modules/registry.py", line 390, in new
    openerp.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/var/tmp/openupgrade/9.0/server/openerp/modules/loading.py", line 446, in load_modules
    force, status, report, loaded_modules, update_module, models_to_check, upg_registry)
  File "/var/tmp/openupgrade/9.0/server/openerp/modules/loading.py", line 331, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check, upg_registry=upg_registry
  File "/var/tmp/openupgrade/9.0/server/openerp/modules/loading.py", line 178, in load_module_graph
    init_module_models(cr, package.name, models)
  File "/var/tmp/openupgrade/9.0/server/openerp/modules/module.py", line 307, in init_module_models
    obj._auto_end(cr, {'module': module_name})
  File "/var/tmp/openupgrade/9.0/server/openerp/api.py", line 250, in wrapper
    return old_api(self, *args, **kwargs)
  File "/var/tmp/openupgrade/9.0/server/openerp/models.py", line 2762, in _auto_end
    cr.execute('ALTER TABLE "%s" ADD FOREIGN KEY ("%s") REFERENCES "%s" ON DELETE %s' % (t, k, r, d))
  File "/var/tmp/openupgrade/9.0/server/openerp/sql_db.py", line 154, in wrapper
    return f(self, *args, **kwargs)
  File "/var/tmp/openupgrade/9.0/server/openerp/sql_db.py", line 233, in execute
    res = self._obj.execute(query, params)
IntegrityError: insert or update on table "hr_applicant" violates foreign key constraint "hr_applicant_source_id_fkey"
DETAIL:  Key (source_id)=(7) is not present in table "utm_source".
```

The migration of "source_id" is done in the post-migration script...
In #682, @pedrobaeza told @cubells that this step can't be done in pre-migration script (because at that time, source_id is pointing to hr_recruitment_source, so referring to utm_source id would raise IntegrityError exception...)

This issue was not there before and I can't see what changed in OpenUpgrade or Odoo or anywhere else...
```
2019-02-14 16:14:24,746 625 INFO database_8_migrated openerp.modules.migration: module hr_recruitment: Running migration [>9.0.1.0] pre-migration
2019-02-14 16:14:24,747 625 INFO database_8_migrated openerp.modules.migration: module hr_recruitment: Running migration [>9.0.1.0] pre-migration
2019-02-14 16:14:24,748 625 INFO database_8_migrated OpenUpgrade: hr_recruitment: pre-migration script called with version 8.0.1.0
2019-02-14 16:14:24,749 625 INFO database_8_migrated OpenUpgrade: table hr_applicant, column priority: renaming to openupgrade_legacy_9_0_priority
2019-02-14 16:14:24,750 625 INFO database_8_migrated OpenUpgrade: table hr_applicant, column availability: renaming to openupgrade_legacy_9_0_availability
2019-02-14 16:14:24,751 625 INFO database_8_migrated OpenUpgrade: table hr_recruitment_source: renaming to openupgrade_legacy_9_0_hr_recruitment_source
2019-02-14 16:14:24,752 625 INFO database_8_migrated OpenUpgrade: table hr_recruitment_source_id_seq: renaming to openupgrade_legacy_9_0_hr_recruitment_source_id_seq
2019-02-14 16:14:25,080 625 INFO database_8_migrated openerp.modules.module: module hr_recruitment: creating or updating database tables
2019-02-14 16:14:26,216 625 INFO database_8_migrated openerp.modules.loading: loading hr_recruitment/security/hr_recruitment_security.xml
2019-02-14 16:14:26,247 625 INFO database_8_migrated openerp.modules.loading: loading hr_recruitment/security/ir.model.access.csv
```
Should we rename column "source_id" to legacy name in pre migration script and remap it during post migration one ?

**Comment by vikas-vg** [comment]
@metaminux did you find any solution for the above problem ?
I'm also getting above issue.

**Comment by vikas-vg** [comment]
Found the workaround to this issue, what I did is update the data file in **utm** module, and add the additional data which is present in table hr_recruitment_source to the table utm_source.

**Comment by alexis-via** [comment]
I confirm this bug.

Let me just re-explain the problem with pointers to the source code:

In v8, the object hr.applicant has a field source_id defined here: https://github.com/odoo/odoo/blob/8.0/addons/hr_recruitment/hr_recruitment.py#L221 It is a M2O to hr.recruitment.source

In v9, the object hr.applicant inherits from utm.mixin (cf https://github.com/odoo/odoo/blob/9.0/addons/hr_recruitment/models/hr_recruitment.py#L84) and the object utm.mixin has a source_id field which is a M2O to utm.source cf https://github.com/odoo/odoo/blob/9.0/addons/utm/models/utm.py#L48

So, the source_id field exists in v8 AND v9, but it points to different objects.

The pre-migration script doesn't do anything special for the source_id field, cf https://github.com/OCA/OpenUpgrade/blob/9.0/addons/hr_recruitment/migrations/9.0.1.0/pre-migration.py

So, when you migrate from v8 to v9, Odoo will fail while installing the new foreign key constraint that point to utm.source:

```
2020-01-20 22:54:55,823 3523 INFO test1 openerp.sql_db: bad query: ALTER TABLE "hr_applicant" ADD FOREIGN KEY ("source_id") REFERENCES "utm_source" ON DELETE set null
2020-01-20 22:54:55,824 3523 ERROR test1 openerp.modules.registry: Failed to load registry
[...]
ForeignKeyViolation: insert or update on table "hr_applicant" violates foreign key constraint "hr_applicant_source_id_fkey"
DETAIL:  Key (source_id)=(6) is not present in table "utm_source".
```

So we need to do something in pre-migration script and post-migration script. I'm not very familiar with openupgrade, but I'll try to make a patch.

**Comment by pedrobaeza** [comment]
Thanks for the insights. The easiest way to go is to rename the column in pre, but you'll lose the possible sources. To be completed, you have to do that and also:
- in pre, add a column in utm_source for linking to old hr.recruitment.source records.
- in pre, insert in utm_source tables the equivalent entries for hr.recruitment.source with the id of the old record.
- In post, update hr_applicant with the new utm_source entry linked from the old record ID and its old column.

**Comment by alexis-via** [comment]
OK, I made a patch. I'll create the PR once my migration is done, so that I can check the data in the web interface (I can't view the data in the web interface because I have other problems in this migration).

**Comment by pedrobaeza** [comment]
Fixed in #2168 and improved in #2214
